### PR TITLE
FriCAS spkg-configure and Feature

### LINKED ISSUE #33575: Feature for fricas
<div id="comment:0"></div>

CC:  @orlitzky @tornaria @seblabbe

Component: **interfaces**

Keywords: **FriCAS**

Branch/Commit: **[u/mkoeppe/feature_for_axiom](https://github.com/sagemath/sagetrac-mirror/tree/u/mkoeppe/feature_for_axiom) @ [`73ad696`](https://github.com/sagemath/sagetrac-mirror/commit/73ad696e35101a24c6c5b03adb518fad8ef3a91b)**

_Issue created by migration from https://trac.sagemath.org/ticket/33575_

### LINKED ISSUE #35837: accept fricas from the system
### Is there an existing issue for this?

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.

### Problem Description

system-wide fricas is sometimes avaliable, and may be used instead of Sage-built one. 

### Proposed Solution

create spkg-configure for it

### PR DESCRIPTION
**Pull request by dimpase**
As FriCAS is only used as an executable, this is straightforward; quite a number of systems has Fricas 1.3.8, so this is useful, too.

The FriCAS pexpect interface now uses the new `Executable` feature to determine the absolute pathname of the fricas executable (unless executed remotely). 
This is made possible by a simple refactor of the `sage.interfaces.expect.Expect` class: Computing the effective command line is no longer done in `set_server_and_command` (called by `__init__`); it is delayed until an interface is started and needs the command line.

- Fixes #35837 
- Fixes #33575

Dependencies: The changes in `sage.interfaces` outside of `.expect` and `.fricas` are all from the following PR and do not need review.

- Depends on #36656 (merged here)

### CODE REVIEW

**Comment by mkoeppe** [comment]
The doctests marked `# optional - fricas` won't be run unless you also do #33575

File: build/pkgs/fricas/distros/gentoo.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+sci-mathematics/fricas[sbcl]
```
Comment (kiwifb) [inline build/pkgs/fricas/distros/gentoo.txt]: Why do you impose it to be build with `sbcl`. Sure it is probably the fastest lisp around, but is there a specific reason that you need to have it?

**Comment by dimpase** [comment]
FriCAS people say using SBCL instead of ECL gives 3 to 4 times faster
system.

**Comment by kiwifb** [comment]
I am not saying to make it ECL, I am saying that lips user on Gentoo make their own choices unless there is a technical reason (as in a breaking reason) to do so. If they are interested in speed they will make that choice themselves.

**Comment by dimpase** [comment]
> The doctests marked `# optional - fricas` won't be run unless you also do #33575

This looks quite heavy-handed. Why can't this be done at `./configure` time? What's the magic done by building `fricas` spkg so that #33575 is not needed?

File: build/pkgs/fricas/distros/gentoo.txt (hunk 1 above)
Comment (dimpase) [reply build/pkgs/fricas/distros/gentoo.txt]: ok

**Comment by dimpase** [comment]
OK, I'm trying now the patch https://github.com/sagemath/sagetrac-mirror/commit/73ad696e35101a24c6c5b03adb518fad8ef3a91b.diff from #33575 (it applies with a fizz)

**Comment by dimpase** [comment]
OK, feature seems to be implemented now, please have a look

File: build/pkgs/fricas/spkg-configure.m4 [hunk 2]
```diff
@@ -0,0 +1,16 @@
+SAGE_SPKG_CONFIGURE(
+    [fricas], [
+        AC_CACHE_CHECK([for FriCAS >= 1.3.8], [ac_cv_path_FRICAS], [
```
Comment (mkoeppe) [inline build/pkgs/fricas/spkg-configure.m4:3]: the indentation is a bit wild

**Comment by mkoeppe** [comment]
Has using system fricas been tested on more than 1 machine?

**Comment by dimpase** [comment]
> Has using system fricas been tested on more than 1 machine?

on 2 different machines, 3 different configurations (using ECL as well as using SBCL as lisp compiler)

**Comment by mkoeppe** [comment]
Build&Test fails with lots of errors originating from the feature thing

**Comment by dimpase** [comment]
works for me locally, it's probably unrelated.

unless these errors come from unavailable fricas.

**Comment by mkoeppe** [comment]
Yes, in the test run on GH Actions there is no FriCAS because it's an optional package...

**Comment by dimpase** [comment]
> Build&Test fails with lots of errors originating from the feature thing

well, indeed - I suppose it's due to #33575 pointing to an outdated design, compared to e.g. what's in 
`src/sage/features/csdp.py`

Another potential problem is `is_FriCASElement` - not sure what should come to replace it.

**Comment by mkoeppe** [comment]
> Another potential problem is `is_FriCASElement` - not sure what should come to replace it.

sage.interfaces.abc

**Comment by mkoeppe** [comment]
I'll take a look what needs to be fixed in the feature definition later today.

**Comment by dimpase** [comment]
> I'll take a look what needs to be fixed in the feature definition later today.

I'm actually working on this now. I'll post an update in an hour or two

**Comment by dimpase** [comment]
here is a non-working update. I'm lost at what `options` are supposed to be there.

`sage.features` an unpleasantly fragile kludge. Is it supposed to test whether pexpect FriCAS interface works?
Whether `fricas` executable is there? Really, no idea, no docs...

```
...
FileNotFoundError: [Errno 2] No such file or directory: 'fricas -nosman'

----------------------------------------------------------------------
sage -t --random-seed=195934185104556495832583672009163086833 src/sage/features/fricas.py  # FileNotFoundError in doctesting framework
```

**Comment by mkoeppe** [comment]
The documentation is being added in this PR, no?

File: src/sage/features/fricas.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+        EXAMPLES::
+
+            sage: from sage.features.fricas import FriCAS
+            sage: FriCAS().is_functional()  # optional - fricas
+            FeatureTestResult('fricas', True)
+        """
+        command = ["fricas -nosman", "2+2"]
+        try:
+            lines = subprocess.check_output(command, stderr=subprocess.STDOUT)
```
Comment (mkoeppe) [inline src/sage/features/fricas.py]: if you don't use `shell=True`, this cannot work

**Comment by dimpase** [comment]
Do we need `_install_hints` for `FriCAS` ? It's not like the other packages for which it's provided, one can install it from Sage.

**Comment by dimpase** [comment]
If FriCAS is not available (I just remove the executable), I get doctest errors like
```
File "src/sage/interfaces/fricas.py", line 899, in sage.interfaces.fricas.FriCAS.__reduce__
Failed example:
    FriCAS().__reduce__()
Exception raised:
    Traceback (most recent call last):
      File "/mnt/opt/Sage/sage-dev/src/sage/doctest/forker.py", line 696, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/mnt/opt/Sage/sage-dev/src/sage/doctest/forker.py", line 1106, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.interfaces.fricas.FriCAS.__reduce__[0]>", line 1, in <module>
        FriCAS().__reduce__()
        ^^^^^^^^
      File "sage/misc/lazy_import.pyx", line 404, in sage.misc.lazy_import.LazyImport.__call__
        sage: is_prime(12) == my_isprime(12)
      File "sage/misc/lazy_import.pyx", line 225, in sage.misc.lazy_import.LazyImport.get_object
        return self._get_object()
      File "sage/misc/lazy_import.pyx", line 261, in sage.misc.lazy_import.LazyImport._get_object
        try:
      File "/mnt/opt/Sage/sage-dev/src/sage/interfaces/fricas.py", line 2134, in <module>
        fricas = FriCAS()
                 ^^^^^^^^
      File "/mnt/opt/Sage/sage-dev/src/sage/interfaces/fricas.py", line 293, in __init__
        FriCAS().require()
      File "/mnt/opt/Sage/sage-dev/src/sage/features/__init__.py", line 243, in require
        raise FeatureNotPresentError(self, presence.reason, presence.resolution)
    sage.features.FeatureNotPresentError: fricas is not available.
    Executable 'fricas' not found on PATH.
    To install fricas using the gentoo package manager, you can try to run:
    !sudo emerge sci-mathematics/fricas
    To install fricas using the Sage package manager, you can try to run:
      !sage -i fricas
    No equivalent system packages for pip are known to Sage.
    Further installation instructions might be available at https://fricas.github.io.
```

no idea what's wrong. Could it be due to `./configure --enable-fricas=yes` ?

The branch seems to work (modulo a number of maths doctest errors) if FriCAS is available.

File: src/sage/features/fricas.py (hunk 3 above)
Comment (dimpase) [reply src/sage/features/fricas.py]: I still don't understand what we are doing here. FriCAS feature is something configured; the testing framework should **test the configuration**, and do  not engage in these run-time double-checks.

As soon as testing framework knows that FriCAS should be available, it must just ignore `# optional - fricas` tags,
and treat FriCAS as available.

I don't get why distributions have any need for these, either.

**Comment by kiwifb** [comment]
Configure time is a feature of the sage distribution. sagemath the python package should be a good python citizen and check things at runtime. We cannot rely on the presence of a configuration or install log in the general case.

Stopping to rely on sage-the-distribution is the whole point of `features`.

**Comment by dimpase** [comment]
> We cannot rely on the presence of a configuration

huh? You're a distro, not a random mashup of Python sources, you know what's installed, and what's not, no?

And Pythonic way would to make `sage.interfaces.fricas` into a pip-installable package, and use
[pipenv](https://packaging.python.org/en/latest/key_projects/#pipenv ), not to roll one's own badly designed 
replacement.

**Comment by kiwifb** [comment]
A separate package is a good idea and we'll get to it I am sure. Yes, I am a distro, but I will not teach sage to do something that work only for my distro. That's not sustainable.

**Comment by dimpase** [comment]
anyhow - anyone familiar with sage.feature please take over from here, I can't fix it.

**Comment by mkoeppe** [comment]
rebased

**Comment by mkoeppe** [comment]
@dimpase There wasn't really anything wrong here, but I've made sage.interfaces.fricas a bit prettier.

**Comment by mkoeppe** [comment]
Let's wait for the "CI Linux incremental" runs.

**Comment by mkoeppe** [comment]
OK, I see what's wrong. Eager use of the feature in `FriCAS.__init__`.

**Comment by mkoeppe** [comment]
I'm working on a solution for this, also suitable for other interfaces.

**Comment by dimpase** [comment]
> A separate package is a good idea and we'll get to it I am sure. Yes, I am a distro, but I will not teach sage to do something that work only for my distro. That's not sustainable.

I don't understand this. Checking that FriCAS is installed on Gentoo need not be done via Sage,and I don't see why it's Sage's business to tell you that FriCAS is installed...

**Comment by mkoeppe** [comment]
Responding to https://github.com/sagemath/sage/pull/35838#issuecomment-1616502001:
> > We cannot rely on the presence of a configuration
> 
> huh? You're a distro, not a random mashup of Python sources, you know what's installed, and what's not, no?
> 
> And Pythonic way would to make `sage.interfaces.fricas` into a pip-installable package, and use [pipenv](https://packaging.python.org/en/latest/key_projects/#pipenv), not to roll one's own badly designed replacement.

There's nothing gained from making a pip-installable package consisting of only `sage.interfaces.fricas`; I'm not planning to create such a package in the modularization effort (if this is what François was referring to).

Not sure what pipenv has to do with any of this, but if you imagine that it can install a non-Python dependency (fricas) somehow, no, that's not what it does.

**Comment by mkoeppe** [comment]
> Do we need `_install_hints` for `FriCAS` ? It's not like the other packages for which it's provided, one can install it from Sage.

No, it's not needed because the `spkg` parameter takes care of all of that already.

**Comment by mkoeppe** [comment]
Usually not a good idea to do a rebase that includes the dependency PR.

**Comment by dimpase** [comment]
All good, it seems. 
Tested with system fricas built with sbcl, as well as fricas built with ECL (both on Gentoo - @orlitzky )
Also tests fine without fricas.

 I can't officially review this, so, please...

**Review by mkoeppe — APPROVED** [review]

**Comment by dimpase** [comment]
@mantepse  - this allows using system-wide FriCAS.

**Comment by dimpase** [comment]
fricas 1.3.10 is meanwhile out.

**Pull request merged**